src/ImageIO/CGImageMetadata.cs

```diff
@@ -118,15 +118,52 @@ public CGImageMetadata (NSData data)
 			return (result == IntPtr.Zero) ? null : new CGImageMetadataTag (result, true);
 		}
 
+#if NET
+		[DllImport (Constants.ImageIOLibrary)]
+		extern unsafe static void CGImageMetadataEnumerateTagsUsingBlock (/* CGImageMetadataRef __nonnull */ IntPtr metadata,
+			/* CFStringRef __nullable */ IntPtr rootPath, /* CFDictionaryRef __nullable */ IntPtr options,
+			/* __nonnull */ delegate* unmanaged<NativeHandle, NativeHandle, bool> block);
+#else
 		[DllImport (Constants.ImageIOLibrary)]
 		extern static void CGImageMetadataEnumerateTagsUsingBlock (/* CGImageMetadataRef __nonnull */ IntPtr metadata,
 			/* CFStringRef __nullable */ IntPtr rootPath, /* CFDictionaryRef __nullable */ IntPtr options,
 			/* __nonnull */ CGImageMetadataTagBlock block);
+#endif
+
+#if NET
+		static CGImageMetadataTagBlock? callbackBlock;
+		static object callbackLock = new object ();
+
+		[UnmanagedCallersOnly]
+		static bool TagEnumerator (NativeHandle key, NativeHandle value)
+		{
+			if (callbackBlock is null)
+				return false;
+			var nsKey = Runtime.GetNSObject<NSString> (key, true)!;
+			var nsValue = Runtime.GetINativeObject<CGImageMetadataTag> (value, true)!;
+			return callbackBlock (nsKey, nsValue);	
+		}
+#endif
 
 		public void EnumerateTags (NSString? rootPath, CGImageMetadataEnumerateOptions? options, CGImageMetadataTagBlock block)
```

Comment:
You could also provide an overload that works recursively by exposing the function pointer parameter:

```cs
public unsafe void EnumerateTags (NSString? rootPath, CGImageMetadataEnumerateOptions? options, delegate* unmanaged<NativeHandle, NativeHandle, bool> block)
{
    // ...
}
```

and maybe even remove the other overload in the future:

```cs
#if !XAMCORE_5_0
public void EnumerateTags (NSString? rootPath, CGImageMetadataEnumerateOptions? options, CGImageMetadataTagBlock block)
{
    // ...
}
#endif // !XAMCORE_5_0
```

Reply:
That's an interesting idea, but I don't think that exposing an internal unsafe low-level implementation is the best idea either, especially since the nature of the API is one that forces an problematic solution anyway. I would be more inclined to make an overload that returns a list of key value pairs.

Reply:
Actually, I think the whole binding is wrong (and we don't have any tests for it either). The callback isn't a C-style function, it's an Objective-C block, and that makes the whole problem go away, because Objective-C blocks don't have this particular problem.

This means I think the first step should be to write a test for the API :)